sql/core/src/main/scala/org/apache/spark/sql/Column.scala

```diff
@@ -133,6 +133,15 @@ class Column(protected[sql] val expr: Expression) extends Logging {
 
     case func: UnresolvedFunction => UnresolvedAlias(func, Some(func.prettyString))
 
+    // If we have a top level Cast, there is a chance to give it a better alias, if there is a
+    // NamedExpression under this Cast.
+    case c: Cast => c.transformUp {
+      case Cast(ne: NamedExpression, to) => UnresolvedAlias(Cast(ne, to))
+    } match {
+      case ne: NamedExpression => ne
+      case other => Alias(expr, expr.prettyString)()
+    }
```

Comment:
We use `transformUp` at here to make this logic equivalent with the original logic that we have in the `cast`, right (since we want to fix it in 1.6)? (in future, will it be possible to just use `cast Cast(ne: NamedExpression, to) => UnresolvedAlias(Cast(ne, to))`?)